gflownet/envs/tree.py

```diff
@@ -0,0 +1,1079 @@
+            action is valid.
+
+        Returns
+        -------
+        self.state : list
+            The sequence after executing the action
+
+        action : tuple
+            Action executed
+
+        valid : bool
+            False, if the action is not allowed for the current state.
+        """
+        # Replace the continuous value of threshold by -1 to allow checking it
+        action_to_check = self.action2representative(action)
+        do_step, self.state, action_to_check = self._pre_step(
+            action_to_check, skip_mask_check or self.skip_mask_check
+        )
+        if not do_step:
+            return self.state, action, False
+
+        self.n_actions += 1
+
+        if action != self.eos:
+            action_type, action_value = action
+
+            if action_type == ActionType.PICK_LEAF:
+                self._pick_leaf(action_value)
+            else:
+                active_node = self._find_active(self.state)
+                if action_type == ActionType.PICK_FEATURE:
+                    self._pick_feature(active_node, action_value)
+                elif action_type == ActionType.PICK_THRESHOLD:
+                    self._pick_threshold(active_node, action_value)
+                elif action_type == ActionType.PICK_OPERATOR:
+                    self._pick_operator(active_node, action_value)
+                else:
+                    raise NotImplementedError(
+                        f"Unrecognized action type: {action_type}."
+                    )
+
+            return self.state, action, True
+        else:
+            self.done = True
+            return self.state, action, True
+
+    def set_state(self, state: List, done: Optional[bool] = False):
+        """
+        Sets the state and done. If done is True but incompatible with state (Stage is
+        not COMPLETE), then force done False and print warning.
+        """
+        if done is True and self._get_stage() != Stage.COMPLETE:
+            done = False
+            warnings.warn(
+                f"""
+            Attempted to set state {self.state2readable(state)} with done = True, which
+            is not compatible with the environment. Forcing done = False.
+            """
+            )
+        return super().set_state(state, done)
+
+    def sample_actions(
+        self,
+        policy_outputs: TensorType["n_states", "policy_output_dim"],
+        sampling_method: str = "policy",
+        mask_invalid_actions: TensorType["n_states", "action_space_dim"] = None,
+        temperature_logits: float = 1.0,
+    ) -> Tuple[List[Tuple], TensorType["n_states"]]:
+        """
+        Samples a batch of actions from a batch of policy outputs.
+        """
+        n_states = policy_outputs.shape[0]
+        logprobs = torch.zeros(n_states, device=self.device, dtype=self.float)
+        # Discrete actions
+        mask_discrete = mask_invalid_actions[:, self._action_index_pick_threshold]
```

Comment:
I might be missing something, but aren't the actions associated with picking threshold continuous, not discrete? So shouldn't the mask associated with discrete actions be *everything but* `[:, self._action_index_pick_threshold]`?

Reply:
```
mask_discrete = mask_invalid_actions[:, self._action_index_pick_threshold]
```
What this line does is creating a boolean mask of the elements (actions) in the batch that are discrete. It is done by taking the value (True or False) in the mask of invalid actions corresponding to the representative of the action "pick threshold", indicated by `self._action_index_pick_threshold`. If it is True, then the action "pick threshold" is invalid and therefore the action must be discrete and it will become `True` in `mask_discrete`. Does that make sense?

Reply:
Oh, I see - makes sense!